### PR TITLE
Pathwise sampling

### PR DESCRIPTION
**Pull request by rossviljoen**
This PR pretty much just implements Matheron's rule from [1]. It defines a function:

`pathwise_sample(f::ApproxPosteriorGP{<:SVGP}, x::AbstractVector, prior_sample_function, num_samples)`

which uses whatever `prior_sample_function` you give it to take an (approximate) sample from the prior. Then, it takes a sample from `q(u)` and updates the prior sample to obtain a posterior sample.

`prior_sample_function` can be anything that returns a sample from the prior - currently the only implementation is `rand`, but other methods (e.g. Random Fourier Features, Karhunen–Loève expansions) can be implemented in other packages.

#7 

[1] https://arxiv.org/abs/2002.09309

---
TODO:

- [ ] tests

### CODE REVIEW

File: src/svgp.jl [hunk 1]
```diff
@@ -77,6 +77,72 @@ function StatsBase.mean_and_var(f::ApproxPosteriorGP{<:SVGP}, x::AbstractVector)
...
+`prior_sample_function` can be any function which returns exact or approximate
+samples from the prior. It must have the signature:
+
+`prior_sample_function(rng::AbstractRNG, fx::FiniteGP, num_samples::Int)`
+
+[2] - James T. Wilson and Viacheslav Borovitskiy and Alexander Terenin and Peter
+Mostowsky and Marc Peter Deisenroth. "Efficiently Sampling Functions from
+Gaussian Process Posteriors" ICML, 2020.
+"""
+function pathwise_sample(
```
Comment (willtebbutt) [inline src/svgp.jl:113]: In terms of design, it seems a shame to require that the locations at which the sample it to be evaluated are needed when the sample is drawn. I wonder whether it would be doable to make this function return a function that can be evaluated at any new input / collection of inputs? So the signature would become something like
```julia
struct ApproxPostSample
    
end

(f::ApproxPostSample)(x) = prior(x) + update(x)

# Optimisation for multiple prediction locations.
map(f::ApproxPostsample, x::AbstractVector) = map(prior, x) + map(update, x)

function pathwise_sample(
    rng::AbstractRNG,
    f::AbstractPosteriorGP{<:SVGP},
    prior_sample_function;
    num_samples=1::Int,
)
    ...
    return ApproxPostSample(prior_sample, terms_needed_for_posterior_update...)
end
```

Comment (rossviljoen) [reply src/svgp.jl:113]: That's more or less the original version I had - correct me if I'm wrong but isn't sampling functions like that only possible if you use some kind of a weight space approximation? (i.e you wouldn't be able to do it if you were just using `rand` right?) So, I wrote it this way to allow you to use literally any kind of prior sample function.

I agree though - perhaps that's too general to the point of uselessness. It might then be better to have this function in the RFF package/whatever instead?

Comment (willtebbutt) [reply src/svgp.jl:113]: So I guess what I'm saying is that, in my opinion, your original approach is something that would be cool to have in this package, even if we also have to define an RFF thing.

Comment (rossviljoen) [reply src/svgp.jl:113]: > I worry that might be the case -- it's unclear to me that, under the current API, much is added over just writing rand(f(x)) or whatever. Or is this not an accurate reflection of the benefits?

I think it's actually strictly worse if you don't use an approximation :smile: - you're right, there's not much point in doing it this way, I'll change it to a function sample version.

**Comment by st--** [comment]
@rossviljoen the lack of tests and the resulting codecov comments make the Github diff hard to read, it looks like you can exclude lines and code blocks in Coverage.jl (see https://github.com/JuliaCI/Coverage.jl/issues/326), how about wrapping the new file in ```julia
# COV_EXCL_START
...
# COV_EXCL_STOP
```
for now?

File: src/svgp.jl [hunk 2]
```diff
@@ -79,5 +79,47 @@ end
 
 inducing_points(f::ApproxPosteriorGP{<:SVGP}) = f.approx.fz.x
 
-_chol_cov(q::AbstractMvNormal) = cholesky(Symmetric(cov(q)))
-_chol_cov(q::MvNormal) = cholesky(q.Σ)
+## Pathwise posterior sampling
+
+struct PosteriorFunctionSamples
```
Comment (willtebbutt) [inline src/svgp.jl:84]: Could this be made type-stable please?

File: src/svgp.jl [hunk 3]
```diff
@@ -79,5 +79,47 @@ end
...
+    num_samples
+    prior  # prior(x) returns an approx prior sample
+    update  # update(x) computes the required pathwise update
+end
+
+(f::PosteriorFunctionSamples)(x) = f([x])
+(f::PosteriorFunctionSamples)(x::AbstractVector) = f.prior(x) + f.update(x)
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/svgp.jl:91]: Do we have a notion of `output_dims`? i.e. since we take the extended-input-domain view of MOGPs throughout the ecosystem, shouldn't it always be possible to think in terms of a matrix of size `(num_points, num_samples)`?

> Alternatively, pathwise_sample could return a vector of functions (one per sample)?

My inclination would be not to define this function like this, but to instead do something like
```julia
map(f::PosteriorFunctionSamples, x::AbstractVector) = map(f.prior, x) + map(f.update, x)
```
You could then represent the output with a `ColVecs` or something.

File: src/svgp.jl [hunk 4]
```diff
@@ -79,5 +79,47 @@ end
...
+# TODO: docstrings
+function pathwise_sample(
+    rng::AbstractRNG,
+    f::ApproxPosteriorGP{<:SVGP},
+    prior_sample_function; # TODO: better name?
+    num_samples=1::Int,
+)
+    svgp = f.approx
+    z = svgp.fz.x
+    input_dims = length(z[1])  # TODO: what's the best way of getting this?
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/svgp.jl:102]: I've been thinking about this. My thinking is that we need to introduce a new kernel type somewhere that wraps an existing kernel and gives us information about the dimension of input that we expect it to apply to -- probably we want to add this in KernelFunctions.

Something like
```julia
# Euclidean domain with D dimensions.
struct Euclidean
    D::Int
end

dim(domain::Euclidean) = domain.D

"""
    DomainKernel{Tkernel<:Kernel, Tdomain}

A kernel with additional information specifying the domain of the kernel.
This provides a complete specification of a kernel, whereas most kernels by themselves do
not provide the domain information necessary to construct e.g. Random Fourier Feature
approximations to themselves.
"""
struct DomainKernel{Tkernel<:Kernel, Tdomain} <: Kernel
    kernel::Tkernel
    domain::Tdomain
end

kernelmatrix(k::DomainKernel, x::AbstractVector) = kernelmatrix(k.kernel, x)

function kernelmatrix(k::DomainKernel, x::AbstractVector, x′::AbstractVector)
    return kernelmatrix(k.kernel, x, x′)
end

kernelmatrix_diag(k::DomainKernel, x::AbstractVector) = kernelmatrix_diag(k.kernel, x)
```

I've written some of my own RFF code as I'm using pathwise sampling in some of my PhD work -- the conclusion that I came to was that our kernels simply don't provide a sufficient amount of information to construct RFF approximations, and really we ought to be able to construct the approximation just given the kernel.

Equally, we don't want to force users to always provide dimension information, so we need something that doesn't break existing code, but allows users to provide enough information to construct the RFF approximation if that's what they want to do.

**Review by willtebbutt — COMMENTED** [review]
Some more comments on design.

File: src/svgp.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (theogf) [reply src/svgp.jl:105]: Related issue: https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/16
Incidentally involving Alex as well :laughing:

File: src/svgp.jl [hunk 5]
```diff
@@ -79,5 +82,50 @@ end
...
+end
+
+(f::PosteriorFunctionSamples)(x) = f([x])
+(f::PosteriorFunctionSamples)(x::AbstractVector) = f.prior(x) + f.update(x)
+
+# TODO: docstrings
+function pathwise_sample(
+    rng::AbstractRNG,
+    f::ApproxPosteriorGP{<:SVGP},
+    prior_sample_function; # TODO: better name?
```
[the remark this answers is not in the document]
Comment (rossviljoen) [reply src/svgp.jl:100]: The intention is that it's anything that constructs a weight space approximation to the prior GP - doesn't have to be RFF

[the remark this answers is not in the document]
Comment (rossviljoen) [reply src/svgp.jl:100]: So yes - it should only depend on the kernel

File: examples/temp.jl [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+num_samples = 100
+
+x_test = sort(rand(input_dims, 500); dims=2)
+
+feature_dims = 1000
+sample_fn = RandomFourierFeatures.gp_rff_approx
+
+function_samples = ApproximateGPs.pathwise_sample(
+    ap, sample_fn, input_dims, feature_dims; num_samples=num_samples
```
Comment (st--) [inline examples/temp.jl:55]: both `input_dims` and `feature_dims` are only needed to pass it through to `weight_space_approx`: how about instead just passing a closure,
```julia
sample_fn(rng, kernel) = RandomFourierFeatures.gp_rff_approx(rng, kernel, input_dims, feature_dims)
```
?

**Comment by st--** [comment]
I'm finding the code flow quite hard to follow, and I'm wondering if that's because of all the carrying-around of closures which you only seem to need to implement `resample`. Conceptually, a pathwise sample is to the GP as a float to Uniform(0,1): I wouldn't expect to be able to resample() a draw from the sample. Is there so much precomputation that it makes sense to have a separate Sample*r* object on which you can call rand() (and then you get a new PathwiseSampleFunction each time)?

File: src/svgp.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/svgp.jl]: > @willtebbutt you argued yourself against having kernels defining their own domain your(past)self 😆 JuliaGaussianProcesses/KernelFunctions.jl#16 (comment)

Indeed I did. IIRC my main concern there was the effect that it would have on end-users -- I'm less concerned if we're able to provide a default domain that users don't see. (If that wasn't my main concern there, it's fair to say that it's my main concern now 😂 )

**Comment by rossviljoen** [comment]
Another kinda funny way of doing this would be by defining a general weight space GP approximation as a thin wrapper around a BayesianLinearRegressor:
```
struct WeightSpaceApproxGP <: AbstractGP
  ϕ  # feature mapping
  blr::BayesianLinearRegressor
end
```
and then implementing the ability to draw function samples in `BLRs` (maybe want this anyway?).

Then there'd be an API for constructing weight space approximations that you could either use for inference or pathwise sampling

Maybe that's overkill though :smile:

**Comment by st--** [comment]
That's a really neat idea actually!

On Thursday, October 14, 2021, Ross Viljoen ***@***.***>
wrote:
> Another kinda funny way of doing this would be by defining a general
weight space GP approximation as a thin wrapper around a
BayesianLinearRegressor:
>
> struct WeightSpaceApproxGP <: AbstractGP
>
> ϕ # feature mapping
>
> blr::BayesianLinearRegressor
>
> end
>
> and then implementing the ability to draw function samples in BLRs (maybe
want this anyway?).
>
> Then there'd be an API for constructing weight space approximations that
you could either use for inference or pathwise sampling
>
> Maybe that's overkill though 😄
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. <
https://ci4.googleusercontent.com/proxy/rl7V3C71HpBu46Nzyi0DdN55hORtpF8h84sZRWYH1lJ6NM_D4DBFDn4qbD8TP_cAAcPBjcT8u6WthY8xcdfxP4HYs8dlr3t3vMhYaNoznHZrzBjsvFYG_PHe2k2BIXElG-3wJU8Tj9YrAwtuD6As8h3vt0rrQDZFUTYalDJYPiFeIVnlFHrp3E663QfpkaOP30ZlYj5clJHqZs6x8p4TmK_YHiVk_lQpNt3iN6s8Cw=s0-d-e1-ft#https://github.com/notifications/beacon/ABL7FDZ563BKG2BEZKH3FQ3UG43MHA5CNFSM5FXDK6C2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOHA73EQY.gif>

**Comment by rossviljoen** [comment]
Superseded by #112
